<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Dereplicate Bruker MALDI Biotyper spectra</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Dereplicate Bruker MALDI Biotyper
spectra</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(maldipickr)</span></code></pre></div>
<!-- WARNING - This vignette is generated by {fusen} from dev/dereplicate-spectra.Rmd: do not edit by hand -->
<p>Bacterial colony identification with the Bruker MALDI Biotyper is a
high-throughput method with the built-in tools, provided that the
selected bacteria belong to the internal database.</p>
<p>Scientific projects where the number of unknown bacteria is expected
to be high needs reference-free methods to be able to reduce the
redundancy of isolated bacterial colonies, a process called
<em>dereplication</em>.</p>
<p><a href="https://doi.org/10.3389/fmicb.2018.01294">Strejcek <em>et
al.</em> (2018)</a> proposed such a method by processing the spectra and
suggest similarity thresholds between spectra above which spectra, and
therefore the measured bacterial colonies, can be considered identical
at a given taxonomic rank. Their processing procedure is implemented in
the <a href="https://github.com/ClavelLab/maldipickr"><code>{maldipickr}</code></a>
package and illustrated in the following vignette.</p>
<p>In addition, we provide functions to enable the dereplication of
different batches of Bruker MALDI Biotyper runs and combine the results,
in order to be able to delineate the clusters from a common similarity
matrix.</p>
<p>More importantly, we provide a function to select a spectra to be
picked in each cluster, a process called <em>cherry-picking</em>,
depending on external metadata and potential out-groups to be excluded
for the current cherry-picking steps.</p>
<div id="process-bruker-maldi-biotyper-spectra" class="section level1">
<h1>Process Bruker MALDI Biotyper spectra</h1>
<div id="process-from-raw-spectra-to-peak-filtering" class="section level2">
<h2>Process from raw spectra to peak filtering</h2>
<p>From the imported raw data from the Bruker MALDI Biotyper, the
processing of the spectra is based on the original implementation, and
run the following tasks:</p>
<ol style="list-style-type: decimal">
<li>Square-root transformation</li>
<li>Mass range trimming to 4-10 kDa as they were deemed most determinant
by Strejcek et al. (2018)</li>
<li>Signal smoothing using the Savitzky-Golay method and a half window
size of 20</li>
<li>Baseline correction with the SNIP procedure</li>
<li>Normalization by Total Ion Current</li>
<li>Peak detection using the SuperSmoother procedure and with a
signal-to-noise ratio above 3</li>
<li>Peak filtering. This step has been added to discard peaks with a
negative signal-to-noise ratio probably due to being on the edge of the
mass range.</li>
</ol>
<p>The full procedure is illustrated in the example below. While in this
case, all the resulting processed spectra, peaks and final spectra
metadata are stored in-memory, the <a href="https://clavellab.github.io/maldipickr/reference/process_spectra.html"><code>process_spectra()</code></a>
function enables storing these files locally for scalable
high-throughput analyses.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>directory_biotyper_spectra <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="st">&quot;toy-species-spectra&quot;</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co"># Import the six spectra</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>spectra_list <span class="ot">&lt;-</span> <span class="fu">import_biotyper_spectra</span>(directory_biotyper_spectra)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>processed <span class="ot">&lt;-</span> <span class="fu">process_spectra</span>(spectra_list)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Overview of the list architecture that is returned</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#  with the list of processed spectra, peaks identified and the</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#  metadata table</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="fu">str</span>(processed, <span class="at">max.level =</span> <span class="dv">2</span>)</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; List of 3</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;  $ spectra :List of 6</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt;   ..$ species1_G2 :Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;   ..$ species2_E11:Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt;   ..$ species2_E12:Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F7 :Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F8 :Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F9 :Formal class &#39;MassSpectrum&#39; [package &quot;MALDIquant&quot;] with 3 slots</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt;  $ peaks   :List of 6</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt;   ..$ species1_G2 :Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt;   ..$ species2_E11:Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt;   ..$ species2_E12:Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F7 :Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F8 :Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt;   ..$ species3_F9 :Formal class &#39;MassPeaks&#39; [package &quot;MALDIquant&quot;] with 4 slots</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt;  $ metadata: tibble [6 × 3] (S3: tbl_df/tbl/data.frame)</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co"># A detailed view of the metadata with the median signal-to-noise</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#  ratio (SNR) and the number of peaks</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a>processed<span class="sc">$</span>metadata</span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="co">#&gt;   name           SNR peaks</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2   5.09    21</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11  5.54    22</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12  5.63    23</span></span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7   4.89    26</span></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8   5.56    25</span></span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9   5.40    25</span></span></code></pre></div>
</div>
<div id="merge-multiple-processed-spectra" class="section level2">
<h2>Merge multiple processed spectra</h2>
<p>During high-throughput analyses, multiples runs of Bruker MALDI
Biotyper are expected resulting in several batches of spectra to be
processed and compared. While their processing is natively independent,
and could natively be run in parallel, the integration of the batches
for their comparison needs an additional step.</p>
<p>The <a href="https://clavellab.github.io/maldipickr/reference/merge_processed_spectra.html"><code>merge_processed_spectra()</code></a>
function aggregates the processed spectra and bins together the detected
peaks, with a tolerance of <span class="math inline">\(0.002\)</span>
between the average peak values in the bin (see <a href="https://rdrr.io/cran/MALDIquant/man/binPeaks-functions.html"><code>MALDIquant::binPeaks</code></a>),
which translate to a tolerance of 2000 ppm. This binning step results in
a <span class="math inline">\(n\times p\)</span> feature matrix (or
intensity matrix), with <span class="math inline">\(n\)</span> rows for
<span class="math inline">\(n\)</span> processed spectra (peak-less
spectra are discarded) and <span class="math inline">\(p\)</span>
columns for the <span class="math inline">\(p\)</span> peaks masses.</p>
<p>By default, as in the Strejeck et al. (2018) procedure, the intensity
values for spectra with missing peaks are interpolated from the
processed spectra signal. The current function enables the analyst to
decide whether to interpolate the values or leave missing peaks as
<code>NA</code> which would then be converted to an null intensity
value.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>directory_biotyper_spectra <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="st">&quot;toy-species-spectra&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># Import the six spectra</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>spectra_list <span class="ot">&lt;-</span> <span class="fu">import_biotyper_spectra</span>(directory_biotyper_spectra)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>processed <span class="ot">&lt;-</span> <span class="fu">process_spectra</span>(spectra_list)</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co"># Merge the spectra to produce the feature matrix</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>fm <span class="ot">&lt;-</span> <span class="fu">merge_processed_spectra</span>(<span class="fu">list</span>(processed))</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co"># The feature matrix has 6 spectra as rows and</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#  35 peaks as columns</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="fu">dim</span>(fm)</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt; [1]  6 35</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co"># Notice the difference when the interpolation is turned off</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>fm_no_interpolation <span class="ot">&lt;-</span> <span class="fu">merge_processed_spectra</span>(</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>  <span class="fu">list</span>(processed),</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>  <span class="at">interpolate_missing =</span> <span class="cn">FALSE</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>)</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="fu">sum</span>(fm <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># 0</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="fu">sum</span>(fm_no_interpolation <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># 68</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co">#&gt; [1] 68</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a><span class="co"># Multiple runs can be aggregated using list()</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a><span class="co"># Merge the spectra to produce the feature matrix</span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a>fm_all <span class="ot">&lt;-</span> <span class="fu">merge_processed_spectra</span>(<span class="fu">list</span>(processed, processed, processed))</span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="co"># The feature matrix has 3×6=18 spectra as rows and</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a><span class="co">#  35 peaks as columns</span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a><span class="fu">dim</span>(fm_all)</span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a><span class="co">#&gt; [1] 18 35</span></span></code></pre></div>
</div>
<div id="compute-a-similarity-matrix-between-all-processed-spectra-not-included" class="section level2">
<h2>Compute a similarity matrix between all processed spectra (not
included)</h2>
<p>Once all the batches of spectra have been processed together, we can
use a distance metric to evaluate how close the spectra are to one
another. <a href="https://doi.org/10.3389/fmicb.2018.01294">Strejcek
<em>et al.</em> (2018)</a> recommend the <em>cosine</em> metric to
compare the spectra and they use the fast implementation in the <a href="https://cran.r-project.org/package=coop"><code>{coop}</code></a>
package.</p>
<p>While we do not provide specific functions to generate the similarity
matrix, we illustrate below how it can be easily computed. Note that the
feature matrix from <a href="https://clavellab.github.io/maldipickr/reference/merge_processed_spectra.html"><code>merge_processed_spectra()</code></a>
has spectra as rows and peaks values as columns. So to get a similarity
matrix between spectra, either the feature matrix must be transposed or
a dedicated function must be used.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># A. Compute the similarity matrix on the transposed feature matrix</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#   using Pearson correlation coefficient</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>sim_matrix <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">cor</span>(<span class="fu">t</span>(fm), <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># B.1 Install the coop package</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co"># install.packages(&quot;coop&quot;)</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co"># B.2 Compute the similarity matrix on the rows of the feature matrix</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>sim_matrix <span class="ot">&lt;-</span> coop<span class="sc">::</span><span class="fu">tcosine</span>(fm)</span></code></pre></div>
</div>
</div>
<div id="delineate-clusters-of-spectra" class="section level1">
<h1>Delineate clusters of spectra</h1>
<div id="from-a-similarity-matrix" class="section level2">
<h2>From a similarity matrix</h2>
<div id="similarity-to-clusters" class="section level3">
<h3>Similarity to clusters</h3>
<p>When the similarity matrix is computed between all pairs of the
studied spectra, the next step is to delineate clusters of spectra to
dereplicate the measured bacterial colonies.</p>
<p>The <a href="https://clavellab.github.io/maldipickr/reference/delineate_with_similarity.html"><code>delineate_with_similarity()</code></a>
is agnostic of the similarity metric used, whether it is the cosine
metric or the Pearson product moment, provided that a numeric threshold
relevant to the metric used is given, and <em>above which</em> two
spectra would be considered similar.</p>
<p>Indeed, the matrix is essentially an adjacency network, where nodes
are spectra and links exist between spectra only if the similarity
between the spectra is <strong>above</strong> (or <strong>equal
to</strong>) the threshold. This representation allows to infer the
clusters easily. A table summarizes for each spectra, to which cluster
number it was assigned to and the size of the cluster, which is the
total number of spectra in the cluster.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Toy similarity matrix between the six example spectra of</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#  three species. The cosine metric is used and a value of</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#  zero indicates dissimilar spectra and a value of one</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#  indicates identical spectra.</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>cosine_similarity <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="fu">c</span>(</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>    <span class="dv">1</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">0.98</span>, <span class="fl">0.98</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    <span class="fl">0.79</span>, <span class="dv">1</span>, <span class="fl">0.98</span>, <span class="fl">0.79</span>, <span class="fl">0.8</span>, <span class="fl">0.8</span>,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>    <span class="fl">0.77</span>, <span class="fl">0.98</span>, <span class="dv">1</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>,</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>    <span class="fl">0.99</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.99</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="dv">1</span>, <span class="dv">1</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  ),</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">6</span>,</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>      <span class="st">&quot;species1_G2&quot;</span>, <span class="st">&quot;species2_E11&quot;</span>, <span class="st">&quot;species2_E12&quot;</span>,</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>      <span class="st">&quot;species3_F7&quot;</span>, <span class="st">&quot;species3_F8&quot;</span>, <span class="st">&quot;species3_F9&quot;</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>    ),</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>      <span class="st">&quot;species1_G2&quot;</span>, <span class="st">&quot;species2_E11&quot;</span>, <span class="st">&quot;species2_E12&quot;</span>,</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>      <span class="st">&quot;species3_F7&quot;</span>, <span class="st">&quot;species3_F8&quot;</span>, <span class="st">&quot;species3_F9&quot;</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>    )</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  )</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>)</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a><span class="co"># Delineate clusters based on a 0.92 threshold applied</span></span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a><span class="co">#  to the similarity matrix</span></span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="fu">delineate_with_similarity</span>(cosine_similarity, <span class="at">threshold =</span> <span class="fl">0.92</span>)</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a><span class="co">#&gt;   name         membership cluster_size</span></span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt;        &lt;int&gt;</span></span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2           1            4</span></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11          2            2</span></span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12          2            2</span></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7           1            4</span></span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8           1            4</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9           1            4</span></span></code></pre></div>
</div>
<div id="set-a-reference-spectrum-for-each-cluster" class="section level3">
<h3>Set a reference spectrum for each cluster</h3>
<p>Once the table of clusters is generated from the similarity matrix, a
reference spectrum can be assigned to each cluster.</p>
<p>We choose to define high-quality spectra as representative spectra of
the clusters using internal information. That is, representative spectra
have, within their cluster, the highest median signal-to-noise ratio and
then the highest number of detected peaks.</p>
<p>The function <a href="https://clavellab.github.io/maldipickr/reference/set_reference_spectra.html"><code>set_reference_spectra()</code></a>
does not change the order of the cluster table but merely adds an
additional column <code>is_reference</code> to indicate whether the
corresponding spectrum is representative of the cluster.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co"># Import the six spectra and</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>processed <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="st">&quot;toy-species-spectra&quot;</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="fu">import_biotyper_spectra</span>() <span class="sc">%&gt;%</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="fu">process_spectra</span>()</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co"># Toy similarity matrix between the six example spectra of</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#  three species. The cosine metric is used and a value of</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#  zero indicates dissimilar spectra and a value of one</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#  indicates identical spectra.</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>cosine_similarity <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>  <span class="fu">c</span>(</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>    <span class="dv">1</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">0.98</span>, <span class="fl">0.98</span>,</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>    <span class="fl">0.79</span>, <span class="dv">1</span>, <span class="fl">0.98</span>, <span class="fl">0.79</span>, <span class="fl">0.8</span>, <span class="fl">0.8</span>,</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>    <span class="fl">0.77</span>, <span class="fl">0.98</span>, <span class="dv">1</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>,</span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>    <span class="fl">0.99</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.99</span>,</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>,</span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="dv">1</span>, <span class="dv">1</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a>  ),</span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">6</span>,</span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(</span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a>      <span class="st">&quot;species1_G2&quot;</span>, <span class="st">&quot;species2_E11&quot;</span>, <span class="st">&quot;species2_E12&quot;</span>,</span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a>      <span class="st">&quot;species3_F7&quot;</span>, <span class="st">&quot;species3_F8&quot;</span>, <span class="st">&quot;species3_F9&quot;</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a>    ),</span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a>      <span class="st">&quot;species1_G2&quot;</span>, <span class="st">&quot;species2_E11&quot;</span>, <span class="st">&quot;species2_E12&quot;</span>,</span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a>      <span class="st">&quot;species3_F7&quot;</span>, <span class="st">&quot;species3_F8&quot;</span>, <span class="st">&quot;species3_F9&quot;</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a>    )</span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a>  )</span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a>)</span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co"># Delineate clusters based on a 0.92 threshold applied</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="co">#  to the similarity matrix</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">delineate_with_similarity</span>(</span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a>  cosine_similarity,</span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a>  <span class="at">threshold =</span> <span class="fl">0.92</span></span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a>)</span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a></span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a><span class="co"># Set reference spectra with the toy example</span></span>
<span id="cb6-44"><a href="#cb6-44" tabindex="-1"></a><span class="fu">set_reference_spectra</span>(clusters, processed<span class="sc">$</span>metadata)</span>
<span id="cb6-45"><a href="#cb6-45" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 6</span></span>
<span id="cb6-46"><a href="#cb6-46" tabindex="-1"></a><span class="co">#&gt;   name         membership cluster_size   SNR peaks is_reference</span></span>
<span id="cb6-47"><a href="#cb6-47" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;       </span></span>
<span id="cb6-48"><a href="#cb6-48" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2           1            4  5.09    21 FALSE       </span></span>
<span id="cb6-49"><a href="#cb6-49" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11          2            2  5.54    22 FALSE       </span></span>
<span id="cb6-50"><a href="#cb6-50" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12          2            2  5.63    23 TRUE        </span></span>
<span id="cb6-51"><a href="#cb6-51" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7           1            4  4.89    26 FALSE       </span></span>
<span id="cb6-52"><a href="#cb6-52" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8           1            4  5.56    25 TRUE        </span></span>
<span id="cb6-53"><a href="#cb6-53" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9           1            4  5.40    25 FALSE</span></span></code></pre></div>
</div>
</div>
<div id="from-taxonomic-identifications" class="section level2">
<h2>From taxonomic identifications</h2>
<p>An alternative to the similarity matrix approach from the previous
section is to rely on the taxonomic identification of the spectra to
delineate clusters. To do so, we must use the Bruker MALDI Biotyper
report from the Compass software that summarize the identification of
the microorganisms using its internal database. Once the report or
reports are imported (in R using <a href="https://clavellab.github.io/maldipickr/reference/read_biotyper_report.html"><code>read_biotyper_report()</code></a>),
the function <a href="https://clavellab.github.io/maldipickr/reference/delineate_with_identification.html"><code>delineate_with_identification()</code></a>
will group spectra based on their identifications.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>report_unknown <span class="ot">&lt;-</span> <span class="fu">read_biotyper_report</span>(</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;biotyper_unknown.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">delineate_with_identification</span>(report_unknown)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; Generating clusters from single report</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; # A tibble: 4 × 3</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt;   name              membership cluster_size</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;                  &lt;int&gt;        &lt;int&gt;</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt; 1 unknown_isolate_1          2            1</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt; 2 unknown_isolate_2          3            1</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt; 3 unknown_isolate_3          1            2</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt; 4 unknown_isolate_4          1            2</span></span></code></pre></div>
<p>Clusters generated from taxonomic identifications can not use the
function <a href="https://clavellab.github.io/maldipickr/reference/set_reference_spectra.html"><code>set_reference_spectra()</code></a>
as the latter relies on peaks information that is not disclosed in the
Biotyper report.</p>
<p>Therefore, users interested in cherry-picking spectra using taxonomic
identifications should use the <a href="https://clavellab.github.io/maldipickr/reference/pick_spectra.html"><code>pick_spectra()</code></a>
function described below with the combination of the input and output
tables of the <a href="https://clavellab.github.io/maldipickr/reference/delineate_with_identification.html"><code>delineate_with_identification()</code></a>
function to pick for instance spectra with the highest log score (using
<code>criteria_column = &quot;bruker_log&quot;</code>).</p>
</div>
<div id="import-clusters-results-generated-by-spede" class="section level2">
<h2>Import clusters results generated by SPeDE</h2>
<p>Raw spectra can also be processed and clustered by another approach,
named <a href="https://doi.org/10.1128/mSystems.00437-19"><code>SPeDE</code></a>,
developed by Dumolin et al. (2019). The resulting dereplication step
produces a comma separated table. The example below illustrates how to
import this table into R to be consistent with the dereplication table
generated within the <a href="https://github.com/ClavelLab/maldipickr"><code>{maldipickr}</code></a>
package.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Reformat the output from SPeDE table</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co"># https://github.com/LM-UGent/SPeDE</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">import_spede_clusters</span>(</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;spede.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 5</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;   name         membership cluster_size quality is_reference</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;dbl&gt;        &lt;int&gt; &lt;chr&gt;   &lt;lgl&gt;       </span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2           1            1 GREEN   TRUE        </span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11          2            2 ORANGE  FALSE       </span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12          2            2 GREEN   TRUE        </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7           3            1 GREEN   TRUE        </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8           4            2 ORANGE  FALSE       </span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9           4            2 GREEN   TRUE</span></span></code></pre></div>
</div>
</div>
<div id="cherry-pick-bruker-maldi-biotyper-spectra" class="section level1">
<h1>Cherry-pick Bruker MALDI Biotyper spectra</h1>
<p>When isolating bacteria from an environment, experimenters want to be
thorough but also work-, time- and cost-savvy. One approach is to reduce
the redundancy of the bacterial isolates by analyzing their MALDI-TOF
spectra from the Bruker Biotyper. All the steps previously described in
this vignette consisted of processing the spectra to be able to pick
only non-redundant spectra, using the <a href="https://clavellab.github.io/maldipickr/reference/pick_spectra.html"><code>pick_spectra()</code></a>
function.</p>
<p>The function, as illustrated in the examples below, can pick spectra
using different types of inputs:</p>
<ul>
<li>the reference spectra information that is present in the cluster
table (after using <a href="https://clavellab.github.io/maldipickr/reference/delineate_with_similarity.html"><code>delineate_with_similarity()</code></a>
or <a href="https://clavellab.github.io/maldipickr/reference/import_spede_clusters.html"><code>import_spede_clusters()</code></a>
functions; see example 1)</li>
<li>an external metadata table containing a variable (e.g., optical
density, fluorescence) to be maximized (default) or minimized per
cluster (see example 2)</li>
</ul>
<p>Spectra, and clusters, can also be excluded from the cherry-picking
decision, a procedure termed <em>masking</em> here. We distinguish two
types of mask that are implemented in the <a href="https://clavellab.github.io/maldipickr/reference/pick_spectra.html"><code>pick_spectra()</code></a>
function:</p>
<ul>
<li><strong>soft mask</strong> that discards the spectra only, if they
correspond for instance to low-quality sample, negative control samples
(see example 3)</li>
<li><strong>hard mask</strong> that discards the spectra <em>and their
clusters</em> (see example 4). This is particularly useful if some
spectra have been previously picked. For instance, to exclude colonies
grown and picked 24h after streaking when comparing with colonies grown
for 72h.</li>
</ul>
<p>Advanced users can also provide directly a cluster table with a
custom sort by cluster to accommodate complex design.</p>
<p>Ultimately, the function delivers a table with as many rows as the
cluster table with an additional logical column named
<code>to_pick</code> to indicate whether the colony associated with the
spectra should be picked (<code>TRUE</code>) or not picked
(<code>FALSE</code>).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># 0. Load a toy example of a tibble of clusters created by</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#   the `delineate_with_similarity` function.</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;clusters_tibble.RDS&quot;</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>    <span class="at">package =</span> <span class="st">&quot;maldipickr&quot;</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  )</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>)</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co"># 1. By default and if no other metadata are provided,</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#   the function picks reference spectra for each clusters.</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co"># N.B: The spectra `name` and `to_pick` columns are moved to the left</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co"># only for clarity using the `relocate()` function.</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="fu">pick_spectra</span>(clusters) <span class="sc">%&gt;%</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">relocate</span>(name, to_pick) <span class="co"># only for clarity</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 7</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#&gt;   name         to_pick membership cluster_size   SNR peaks is_reference</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;        &lt;lgl&gt;        &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;       </span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2  FALSE            1            4  5.09    21 FALSE       </span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11 FALSE            2            2  5.54    22 FALSE       </span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12 TRUE             2            2  5.63    23 TRUE        </span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7  FALSE            1            4  4.89    26 FALSE       </span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8  TRUE             1            4  5.56    25 TRUE        </span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9  FALSE            1            4  5.40    25 FALSE</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="co"># 2.1 Simulate OD600 values with uniform distribution</span></span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="co">#  for each of the colonies we measured with</span></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="co">#  the Bruker MALDI Biotyper</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">104</span>)</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">transmute</span>(</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>  clusters,</span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>  <span class="at">name =</span> name, <span class="at">OD600 =</span> <span class="fu">runif</span>(<span class="at">n =</span> <span class="fu">nrow</span>(clusters))</span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a>)</span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a>metadata</span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a><span class="co">#&gt;   name         OD600</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;        &lt;dbl&gt;</span></span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2  0.364</span></span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11 0.772</span></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12 0.735</span></span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7  0.973</span></span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8  0.740</span></span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9  0.201</span></span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a></span>
<span id="cb9-45"><a href="#cb9-45" tabindex="-1"></a><span class="co"># 2.2 Pick the spectra based on the highest</span></span>
<span id="cb9-46"><a href="#cb9-46" tabindex="-1"></a><span class="co">#   OD600 value per cluster</span></span>
<span id="cb9-47"><a href="#cb9-47" tabindex="-1"></a><span class="fu">pick_spectra</span>(clusters, metadata, <span class="st">&quot;OD600&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-48"><a href="#cb9-48" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">relocate</span>(name, to_pick) <span class="co"># only for clarity</span></span>
<span id="cb9-49"><a href="#cb9-49" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span id="cb9-50"><a href="#cb9-50" tabindex="-1"></a><span class="co">#&gt;   name         to_pick membership cluster_size   SNR peaks is_reference OD600</span></span>
<span id="cb9-51"><a href="#cb9-51" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;        &lt;lgl&gt;        &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;        &lt;dbl&gt;</span></span>
<span id="cb9-52"><a href="#cb9-52" tabindex="-1"></a><span class="co">#&gt; 1 species1_G2  FALSE            1            4  5.09    21 FALSE        0.364</span></span>
<span id="cb9-53"><a href="#cb9-53" tabindex="-1"></a><span class="co">#&gt; 2 species2_E11 TRUE             2            2  5.54    22 FALSE        0.772</span></span>
<span id="cb9-54"><a href="#cb9-54" tabindex="-1"></a><span class="co">#&gt; 3 species2_E12 FALSE            2            2  5.63    23 TRUE         0.735</span></span>
<span id="cb9-55"><a href="#cb9-55" tabindex="-1"></a><span class="co">#&gt; 4 species3_F7  TRUE             1            4  4.89    26 FALSE        0.973</span></span>
<span id="cb9-56"><a href="#cb9-56" tabindex="-1"></a><span class="co">#&gt; 5 species3_F8  FALSE            1            4  5.56    25 TRUE         0.740</span></span>
<span id="cb9-57"><a href="#cb9-57" tabindex="-1"></a><span class="co">#&gt; 6 species3_F9  FALSE            1            4  5.40    25 FALSE        0.201</span></span>
<span id="cb9-58"><a href="#cb9-58" tabindex="-1"></a></span>
<span id="cb9-59"><a href="#cb9-59" tabindex="-1"></a><span class="co"># 3.1 Say that the wells on the right side of the plate are</span></span>
<span id="cb9-60"><a href="#cb9-60" tabindex="-1"></a><span class="co">#   used for negative controls and should not be picked.</span></span>
<span id="cb9-61"><a href="#cb9-61" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb9-62"><a href="#cb9-62" tabindex="-1"></a>  <span class="at">well =</span> <span class="fu">gsub</span>(<span class="st">&quot;.*[A-Z]([0-9]{1,2}$)&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">1&quot;</span>, name) <span class="sc">%&gt;%</span></span>
<span id="cb9-63"><a href="#cb9-63" tabindex="-1"></a>    <span class="fu">strtoi</span>(),</span>
<span id="cb9-64"><a href="#cb9-64" tabindex="-1"></a>  <span class="at">is_edge =</span> <span class="fu">is_well_on_edge</span>(</span>
<span id="cb9-65"><a href="#cb9-65" tabindex="-1"></a>    <span class="at">well_number =</span> well, <span class="at">plate_layout =</span> <span class="dv">96</span>, <span class="at">edges =</span> <span class="st">&quot;right&quot;</span></span>
<span id="cb9-66"><a href="#cb9-66" tabindex="-1"></a>  )</span>
<span id="cb9-67"><a href="#cb9-67" tabindex="-1"></a>)</span>
<span id="cb9-68"><a href="#cb9-68" tabindex="-1"></a></span>
<span id="cb9-69"><a href="#cb9-69" tabindex="-1"></a><span class="co"># 3.2 Pick the spectra after discarding (or soft masking)</span></span>
<span id="cb9-70"><a href="#cb9-70" tabindex="-1"></a><span class="co">#   the spectra indicated by the `is_edge` column.</span></span>
<span id="cb9-71"><a href="#cb9-71" tabindex="-1"></a><span class="fu">pick_spectra</span>(clusters, metadata, <span class="st">&quot;OD600&quot;</span>,</span>
<span id="cb9-72"><a href="#cb9-72" tabindex="-1"></a>  <span class="at">soft_mask_column =</span> <span class="st">&quot;is_edge&quot;</span></span>
<span id="cb9-73"><a href="#cb9-73" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb9-74"><a href="#cb9-74" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">relocate</span>(name, to_pick) <span class="co"># only for clarity</span></span>
<span id="cb9-75"><a href="#cb9-75" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 10</span></span>
<span id="cb9-76"><a href="#cb9-76" tabindex="-1"></a><span class="co">#&gt;   name      to_pick membership cluster_size   SNR peaks is_reference OD600  well</span></span>
<span id="cb9-77"><a href="#cb9-77" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;     &lt;lgl&gt;        &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;        &lt;dbl&gt; &lt;int&gt;</span></span>
<span id="cb9-78"><a href="#cb9-78" tabindex="-1"></a><span class="co">#&gt; 1 species1… FALSE            1            4  5.09    21 FALSE        0.364     2</span></span>
<span id="cb9-79"><a href="#cb9-79" tabindex="-1"></a><span class="co">#&gt; 2 species2… TRUE             2            2  5.54    22 FALSE        0.772    11</span></span>
<span id="cb9-80"><a href="#cb9-80" tabindex="-1"></a><span class="co">#&gt; 3 species2… FALSE            2            2  5.63    23 TRUE         0.735    12</span></span>
<span id="cb9-81"><a href="#cb9-81" tabindex="-1"></a><span class="co">#&gt; 4 species3… TRUE             1            4  4.89    26 FALSE        0.973     7</span></span>
<span id="cb9-82"><a href="#cb9-82" tabindex="-1"></a><span class="co">#&gt; 5 species3… FALSE            1            4  5.56    25 TRUE         0.740     8</span></span>
<span id="cb9-83"><a href="#cb9-83" tabindex="-1"></a><span class="co">#&gt; 6 species3… FALSE            1            4  5.40    25 FALSE        0.201     9</span></span>
<span id="cb9-84"><a href="#cb9-84" tabindex="-1"></a><span class="co">#&gt; # ℹ 1 more variable: is_edge &lt;lgl&gt;</span></span>
<span id="cb9-85"><a href="#cb9-85" tabindex="-1"></a></span>
<span id="cb9-86"><a href="#cb9-86" tabindex="-1"></a><span class="co"># 4.1 Say that some spectra were picked before</span></span>
<span id="cb9-87"><a href="#cb9-87" tabindex="-1"></a><span class="co">#   (e.g., in the column F) in a previous experiment.</span></span>
<span id="cb9-88"><a href="#cb9-88" tabindex="-1"></a><span class="co"># We do not want to pick clusters with those spectra</span></span>
<span id="cb9-89"><a href="#cb9-89" tabindex="-1"></a><span class="co">#   included to limit redundancy.</span></span>
<span id="cb9-90"><a href="#cb9-90" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb9-91"><a href="#cb9-91" tabindex="-1"></a>  <span class="at">picked_before =</span> <span class="fu">grepl</span>(<span class="st">&quot;_F&quot;</span>, name)</span>
<span id="cb9-92"><a href="#cb9-92" tabindex="-1"></a>)</span>
<span id="cb9-93"><a href="#cb9-93" tabindex="-1"></a><span class="co"># 4.2 Pick the spectra from clusters without spectra</span></span>
<span id="cb9-94"><a href="#cb9-94" tabindex="-1"></a><span class="co">#   labeled as `picked_before` (hard masking).</span></span>
<span id="cb9-95"><a href="#cb9-95" tabindex="-1"></a><span class="fu">pick_spectra</span>(clusters, metadata, <span class="st">&quot;OD600&quot;</span>,</span>
<span id="cb9-96"><a href="#cb9-96" tabindex="-1"></a>  <span class="at">hard_mask_column =</span> <span class="st">&quot;picked_before&quot;</span></span>
<span id="cb9-97"><a href="#cb9-97" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb9-98"><a href="#cb9-98" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">relocate</span>(name, to_pick) <span class="co"># only for clarity</span></span>
<span id="cb9-99"><a href="#cb9-99" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 11</span></span>
<span id="cb9-100"><a href="#cb9-100" tabindex="-1"></a><span class="co">#&gt;   name      to_pick membership cluster_size   SNR peaks is_reference OD600  well</span></span>
<span id="cb9-101"><a href="#cb9-101" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;     &lt;lgl&gt;        &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;        &lt;dbl&gt; &lt;int&gt;</span></span>
<span id="cb9-102"><a href="#cb9-102" tabindex="-1"></a><span class="co">#&gt; 1 species1… FALSE            1            4  5.09    21 FALSE        0.364     2</span></span>
<span id="cb9-103"><a href="#cb9-103" tabindex="-1"></a><span class="co">#&gt; 2 species2… TRUE             2            2  5.54    22 FALSE        0.772    11</span></span>
<span id="cb9-104"><a href="#cb9-104" tabindex="-1"></a><span class="co">#&gt; 3 species2… FALSE            2            2  5.63    23 TRUE         0.735    12</span></span>
<span id="cb9-105"><a href="#cb9-105" tabindex="-1"></a><span class="co">#&gt; 4 species3… FALSE            1            4  4.89    26 FALSE        0.973     7</span></span>
<span id="cb9-106"><a href="#cb9-106" tabindex="-1"></a><span class="co">#&gt; 5 species3… FALSE            1            4  5.56    25 TRUE         0.740     8</span></span>
<span id="cb9-107"><a href="#cb9-107" tabindex="-1"></a><span class="co">#&gt; 6 species3… FALSE            1            4  5.40    25 FALSE        0.201     9</span></span>
<span id="cb9-108"><a href="#cb9-108" tabindex="-1"></a><span class="co">#&gt; # ℹ 2 more variables: is_edge &lt;lgl&gt;, picked_before &lt;lgl&gt;</span></span></code></pre></div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<ul>
<li>Dumolin C, Aerts M, Verheyde B, Schellaert S, Vandamme T, Van Der
Jeugt F, De Canck E, Cnockaert M, Wieme AD, Cleenwerck I, Peiren J,
Dawyndt P, Vandamme P, &amp; Carlier A. (2019). “Introducing SPeDE:
High-Throughput Dereplication and Accurate Determination of Microbial
Diversity from Matrix-Assisted Laser Desorption–Ionization Time of
Flight Mass Spectrometry Data”. <em>MSystems</em> 4(5). <a href="doi:10.1128/msystems.00437-19" class="uri">doi:10.1128/msystems.00437-19</a>.</li>
<li>Strejcek M, Smrhova T, Junkova P &amp; Uhlik O (2018). “Whole-Cell
MALDI-TOF MS versus 16S rRNA Gene Analysis for Identification and
Dereplication of Recurrent Bacterial Isolates.” <em>Frontiers in
Microbiology</em> 9 <a href="doi:10.3389/fmicb.2018.01294" class="uri">doi:10.3389/fmicb.2018.01294</a>.</li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
